### PR TITLE
Fixed: Improved range by merge delta transformation case.

### LINKED ISSUE #1132: Range#getTransformedByDelta returns incorrect results for uncommon MergeDeltas
A special case transformation of range by `MergeDelta` assumes that `MoveOperation` in `MergeDelta` starts at offset `0`.

This is not true for all `MoveOperation`s in `MergeDelta`. Sometimes, due to transformations, `MergeDelta` may contain `MoveOperation` that does not start from offset `0`.

### PR DESCRIPTION
**Pull request by scofalik**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Fixed a bug in `Range#getTransformedByDelta()` that caused editor crash after some `MergeDelta`s was transformed. Closes #1132.

### CODE REVIEW

File: src/model/range.js [hunk 1]
```diff
@@ -493,7 +493,10 @@ export default class Range {
 				// Without fix, the range would end up in the graveyard, together with removed element.
 				// <p>foo</p><p>[]bar</p> -> <p>foobar</p><p>[]</p> -> <p>foobar</p> -> <p>foo[]bar</p>
 				// <p>foo</p><p>bar[]</p>
-				return [ new Range( targetPosition.getShiftedBy( this.start.offset ) ) ];
+				// return [ new Range( targetPosition.getShiftedBy( this.start.offset ) ) ];
```
Comment (Reinmar) [inline src/model/range.js]: ⬆️

**Comment by Reinmar** [comment]
I hope to understand at least this OT-related PR but you haven't left a comment why is this change needed. How can merge happen on a non-zero index?

File: src/model/range.js [hunk 2]
```diff
@@ -493,7 +493,10 @@ export default class Range {
 				// Without fix, the range would end up in the graveyard, together with removed element.
 				// <p>foo</p><p>[]bar</p> -> <p>foobar</p><p>[]</p> -> <p>foobar</p> -> <p>foo[]bar</p>
 				// <p>foo</p><p>bar[]</p>
```
Comment (Reinmar) [inline src/model/range.js]: What does this comment mean? That last bit?

File: tests/model/range.js [hunk 3]
```diff
@@ -1056,6 +1059,22 @@ describe( 'Range', () => {
 				expect( transformed[ 0 ].start.path ).to.deep.equal( [ 0, 0, 1 ] );
 				expect( transformed[ 0 ].end.path ).to.deep.equal( [ 0, 1, 1 ] );
 			} );
+
+			// #1132
+			it( 'merge delta has move operation that does not start from offset 0', () => {
```
Comment (Reinmar) [inline tests/model/range.js]: Such tests without initial comments like these in lines 1048-1051 are completely unreadable for me (at least ;P)

File: src/model/range.js (hunk 2 above)
Comment (scofalik) [reply src/model/range.js]: It is the same as the above, just the input data is same. I was hoping that the explanation above will be enough to understand that these are similar cases with different input.

Comment (scofalik) [reply src/model/range.js]: Well maybe output is not exactly the same because range ends up in a different spot. I'll write whole example.

**Comment by scofalik** [comment]
> I hope to understand at least this OT-related PR but you haven't left a comment why is this change needed. How can merge happen on a non-zero index?

Yes it is because of OT but the example is complicated so I didn't want to leave long comment about it that would be not understood by anyone who does not know how transformations work.

Basically this happens:
1. `SplitDelta` A is transformed by `SplitDelta` B and they both split in same element, but on different positions. This is a collaborative editing scenario.
2. Both `SplitDelta`s are undone.
3. One of `MoveOperation` in reversed `SplitDelta` (`MergeDelta`) becomes something like:
`move [ a, x ] - [ a, y ] -> [ a, x ]` where `x` is not `0`. Basically, this move operation does not change anything as we move a range to the start of it. But because of special case in `Range` transformation method, it threw.

Now I understand that point 3. raises two concerns:
1. If this move operation does not do anything, why it is not a `NoOperation`?
2. If this move operation does not do anything, why we don't just skip such operations in `Range#getTransformedByDelta()`.

Here are some answers:
1. We already have problems in OT because of transforming operations to `NoOperation` when they seem to not do anything. This is a recurring theme that we already fixed on multiple occasions in multiple transformation cases. For sure we cannot transform to `NoOperation` during the transformation process. Maybe we could transform to `NoOperation` at the end when we already have results and are ready to apply them. But this is maybe.

2. We could just check if move operation does anything relevant and return if not. But at this point I cannot guarantee that there is no way to have such operation generated through OT:
`move [ a, x ] - [ a, y ] -> [ b, z ]` where `x` is not `0`.
So to prevent future errors I wanted to fix how `offset` is calculated because it seemed wrong. The new `offset` formula is same for `sourceRange.start.offset == 0` but is also correct for `sourceRange.start.offset != 0`.

**Comment by scofalik** [comment]
What's up with this PR?

**Comment by Reinmar** [comment]
W8ing for me having time to check it.

**Pull request merged**